### PR TITLE
Adds recursion page

### LINKED ISSUE #54: Add recursion docs

### PR DESCRIPTION
**Pull request by bkase**
Fixes #54

### CODE REVIEW

File: docs/zkapps/recursion.mdx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  methods: {
+    baseCase: {
+      privateInputs: [],
+
+      method(publicInput: Field) {
+        publicInput.assertEquals(Field.zero);
+      },
+    },
+
+    inductiveCase: {
```
Comment (jasongitmail) [inline docs/zkapps/recursion.mdx:83]: Can we use a more accessible (ELI5) method name?

**Review by jasongitmail — COMMENTED** [review]
Left some comments. I got most the way through.

Overall, wondering if we could organize this more clearly. I.e. 1.) Overview to explain what "why" with example use cases and give a high level understanding. 2.) Then maybe we should have an example that shows the basic syntax, then adds a 2nd method, then shows how to run it works recursively. 3.) Then shows how to settle the main by verifying the proof in a method on SmartContract.  

Then I might summarize that that's the basic workflow, but then they can go further with the below.

Then a section about settling to the chain by verifying the proof within a method on SmartContract.

That might feel more organized to me. wdyt?

There are a number of names chosen in here that could probably be made more accessible to a broader audience (i.e. inductiveCase)

File: docs/zkapps/recursion.mdx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+Kimchi, the proof system that backs SnarkyJS, supports arbitrary infinite recursive proof construction of circuits. We can use this to our advantage to verify any zero knowledge program as part of our zkApp.
+
+## Recursively Verifying a simple program within a zkApp
+
+First we need to define a program. Just like in zkApps, arguments to methods are private. This time, we can specify what our public input shape is. This one will be simple and just have one method that proves that the public input is zero:
+
+```typescript
+import { Field, Experimental } from 'snarkyjs';
+
+let SimpleProgram = Experimental.ZkProgram({
```
[the remark this answers is not in the document]
Comment (mitschabaude) [reply docs/zkapps/recursion.mdx:26]: Noooooo

File: docs/zkapps/recursion.mdx [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+Now we can use it to create a proof:
+
+```typescript
+let proof = await SimpleProgram.run(Field.zero);
+```
+
+And we can verify this proof from within any zkApp method:
+
+```typescript
+@method foo() {
```
Comment (mitschabaude) [inline docs/zkapps/recursion.mdx:56]: I think this method should have an argument 'proof: Proof'

**Review by mitschabaude — COMMENTED** [review]
Looks great so far! I would argue that it makes sense to have it in its own section, as we should move towards more detailed docs & more sections anyway

File: docs/zkapps/recursion.mdx (hunk 2 above)
[the remark this answers is not in the document]
Comment (Trivo25) [reply docs/zkapps/recursion.mdx:26]: let's just use `var` and everyone will be happy ;)

File: docs/zkapps/recursion.mdx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+let proof1 = await AddOne.step(Field.one, proof);
+```
+
+And we can repeat this as many times as we want:
+
+```typescript
+let proof2 = await AddOne.step(Field(2), proof1);
+```
+
+Finally we can verify the proof from within a zkApp like we did above:
```
Comment (jasongitmail) [inline docs/zkapps/recursion.mdx:135]: * "from within a SmartContract" correct?

Also, this switches from showing how to call ZkProgram, to showing the implementation of the method in the SmartContract. Should probably show how to call it, or provide consistency for the reader.

File: docs/zkapps/recursion.mdx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+```typescript
+let proof = await SimpleProgram.run(Field.zero);
+```
+
+And we can verify this proof from within any zkApp method:
+
+```typescript
+@method foo() {
+  proof.verify().assertTrue();
+  //... the rest of our method
```
[the remark this answers is not in the document]
Comment (jasongitmail) [reply docs/zkapps/recursion.mdx]: This would still be useful @bkase

**Review by jasongitmail — APPROVED** [review]

File: docs/zkapps/recursion.mdx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+First we compile this program and make the base proof as before:
+
+```typescript
+let { verificationKey } = await AddOne.compile();
+
+let proof = await AddOne.baseCase(Field.zero);
+```
+
+This time we can use this proof as input to recursively add one again:
```
[the remark this answers is not in the document]
Comment (bkase) [reply docs/zkapps/recursion.mdx:123]: this needs to be 2 because it's the "result of adding 1 to (a proof that we've added 1 to zero)"

We intend on fixing this API so you don't need to pass what is semantically the return value as a parameter (and instead can return it!), but this requires some heavy refactoring of the internals of SnarkyJS. I've captured this in an issue because I couldn't find the one that I thought we had https://github.com/o1-labs/snarkyjs/issues/523 and alluded to it in the Note at the top.

**Review by mitschabaude — APPROVED** [review]
lgtm!
